### PR TITLE
AF-1251: Support to experimental features

### PR DESCRIPTION
**Pull request by pefernan**
Adding the experimental CMS data management components 

@porcelli @ederign  could you please take a look?

This is a part of an ensemble, please merge with:
https://github.com/kiegroup/appformer/pull/415
https://github.com/kiegroup/kie-wb-common/pull/1965
https://github.com/kiegroup/kie-wb-distributions/pull/779
https://github.com/kiegroup/drools-wb/pull/887
https://github.com/kiegroup/optaplanner-wb/pull/291

### CODE REVIEW

File: kie-wb-common-forms/kie-wb-common-forms-cms/kie-wb-common-forms-cms-layout-components/kie-wb-common-forms-cms-layout-components-client/src/main/java/org/kie/workbench/common/forms/cms/components/client/ui/DataManagementGroupProvider.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                    DataManagementDragComponent dragComponent = beanDef.getInstance();
+                    group.addLayoutDragComponent(dragComponent.getDragComponentTitle(), dragComponent);
+                });
+
+        return group;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return experimentalFeaturesRegistryService.isFeatureEnabled(DataManagementFeatureDefinition.DATA_MANAGEMENT_FEATURE_KEY);
```
Comment (ederign) [inline kie-wb-common-forms/kie-wb-common-forms-cms/kie-wb-common-forms-cms-layout-components/kie-wb-common-forms-cms-layout-components-client/src/main/java/org/kie/workbench/common/forms/cms/components/client/ui/DataManagementGroupProvider.java]: @porcelli @tomasdavidorg  experimental feature enablement.

**Comment by ederign** [comment]
@tomasdavidorg all the other PR's are on framework level, so we have all the coverage.

This one is the first usage of our experimental features stuff. @porcelli can give you more details about the rules for this kind of PR.

**Comment by pefernan** [comment]
jenkins retest this

**Comment by pefernan** [comment]
jenkins retest this

**Comment by tomasdavidorg** [comment]
As @ederign said. We should have some rules for adding experimental features. It is clear that an experimental feature should not destroy the workbench when the feature is turned off but what about its bugs. Are we ok that the feature is somehow broken and contains these bugs and merged to the master? Should be the bugs reported? Would it be better to give the feature the same level of quality as for regular features? Or we can consider if we would rather use the feature branches as we were planed. We can then avoid big PRs and reviewers could watch the progress

About this concrete PRs. It consists of two parts (framework and the CMS data management components ) which should be rather divided. Even so I can approve the PR even it is quite big and I can see some issues there but firstly we should answer the questions. 

@ederign @porcelli @pefernan WDYT?

**Comment by Rikkola** [comment]
Just github lurking here.

I just tried to push a bad experimental PR in and @manstis stopped me. That got me thinking about what the quality should be for the experimental PRs. In the end I think they should be on the same level as any PRs. Good test coverage, no commented code, no bugs, docs updated and so on. The reason for this is. Once the garbage is in the bar to do something about it goes high.

To prove this. Do a search for "todo" comments in our code base and see what will "soon be fixed". :-)

**Comment by ederign** [comment]
The main goal of this feature is to be able to 'experiment' spikes on our codebase that doesn't exactly match our quality code standards. Those features *must* be hidden by this new Pere mechanism. 

What we are addressing here is to have the ability to try new things and doesn't lose some spike because of the rotten code.

This code should not break anything in production and also we have a clear separation border between the experimental feature and the workbench code.

For instance, we cannot do a refactoring or improve something in production with this 'experimental tag' but can have an experimental implementation of an editor hidden by Pere's mechanism.

For this special cases, we can admit code that doesn't follow our code standards of QE, tests etc. In future, if someone decides to 'promote' this feature (in fact removing Pere's mechanism) this code *needs* to be reviewed again and also approved by QE. I'll make sure that this rule will be followed on AppFormer codebase, and hope that others help us to make this a rule.

What do you guys think? @tomasdavidorg @Rikkola 

Can we go ahead and merge this PR? or @tomasdavidorg do you think this PR will introduce some issue in the workbench?

**Comment by jhrcek** [comment]
Just a comment, don't let this influence merging of this PR :)
I fear that concerns mentioned by @tomasdavidorg and @Rikkola are valid: once we have buggy /  insufficiently tested code merged in, it will be harder to enforce all the quality constraints.
The code and the potential future tests (that will need to be written to promote the feature from experimental) will be separated by time gap, with some other changes included, which will make the review of coverage / feature harder. In any case we should definitely not "misuse" the experimental status as an excuse for "I'll write test for this later". If we can prevent that, than I'm ok with merging this.

**Comment by tomasdavidorg** [comment]
@ederign I agree with @jhrcek. 

I talked with @jhrcek and it would be best if we split this into framework part and data management feature part. It will help us understand the feature more and we can do better PR review. @pefernan Is that possible?

**Comment by pefernan** [comment]
@tomasdavidorg @jhrcek I'm going to split this PR and https://github.com/kiegroup/kie-wb-distributions/pull/779 in two:
- first on for the experimental framework
- second for the experimental feature itself

works for you?

**Comment by tomasdavidorg** [comment]
@pefernan That would be great. Thanks.

**Comment by pefernan** [comment]
@tomasdavidorg @jhrcek done, cleaned up the experimental stuff

**Comment by pefernan** [comment]
jenkins retest this please

**Comment by pefernan** [comment]
Closing

**Pull request closed without merging**